### PR TITLE
pathlib.Path support for keras 

### PR DESCRIPTION
**Pull request by ngc92**
this adds pathlib support to keras functions taking paths as arguments, see #37357

### CODE REVIEW

File: tensorflow/python/keras/saving/save_test.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/saving/save_test.py]: We are still releasing Python3.5 packages. Do we need to increase major to 6?

File: tensorflow/python/keras/utils/io_utils.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/utils/io_utils.py]: I would wrap the `if` with the function instead, hiding the version  switch inside.

```suggestion
def path_to_string(path):
  """Convert a `PathLike` path to a string path, if supported.

  If Python interpreter supports `os.fspath`, then this function works as follows:
  If given a non-string typed path object, converts it to its string
  representation. If the object passed to `path` is not `PathLike` according
  to https://www.python.org/dev/peps/pep-0519 , then the object is returned
  unchanged. This allows e.g. passthrough of file objects through this
  function.
  
  Otherwise,  just return input unchaged, for older versions of
  python that do not support the file system path protocol.

  See https://www.python.org/dev/peps/pep-0519 .
  """
  if sys.version_info >= (3, 6):
    import os
    path = os.fspath(path) if isinstance(path, os.PathLike) else path
    delete os
    return path
  else:
    return path


```

**Comment by mihaimaruseac** [comment]
I think we should mostly do the conversion just before calling the filesystem layer (e.g., in `tf.io.gfile....` methods).

For the cases where those methods are not used and developers used the natural Python functions, I think we should first attempt a conversion to the wrapped ones. Only if this conversion does not make sense should we try to convert path-like objects to string in the other cases.

The utility can reside in TF proper, might need a public `tf.` name (via `tf_export` decorator).

We should also get a review from the API owners

File: tensorflow/python/keras/saving/save_test.py
Comment (ngc92) [reply tensorflow/python/keras/saving/save_test.py]: We could also have three different code paths:
1) Python <= 3.4, where there is no (native) pathlib
2) 3.4, 3.5 They have path lib, but do lack the `os.PathLike`, `os.fspath` functions that allow generic implementations of path objects that is not limited to `pathlib.Path`
3) >=3.6 The filesystem path protocol based implementation above

In theory there would be could a fourth one, based on the pypy pathlib backport to python2, but since support is dropping for it I think we don't need that.

File: tensorflow/python/keras/utils/io_utils.py
Comment (ngc92) [reply tensorflow/python/keras/utils/io_utils.py]: I wanted to keep the amount of new code introduced in the code path minimal, and since the python version cannot change after startup, I thought I'd do the test just once.

Your are right though, that in the case of the keras functions covered here, this is highly unlikely to make any performance difference, and anything that interfaces with files is probably many orders of magnitude slower than the version checks.

Why do you think "if in function" is prefereable to "function in if"?

Comment (mihaimaruseac) [reply tensorflow/python/keras/utils/io_utils.py]: I wanted documentation to be in a single place for all codepaths and also to move the `if` execution from import time to the first time the function gets executed. `import tensorflow` is already slow, so we're trying to do as little as possible at import time.

File: tensorflow/python/keras/saving/save_test.py
Comment (mihaimaruseac) [reply tensorflow/python/keras/saving/save_test.py]: Can we use the utility function here instead of switching on version?

Comment (ngc92) [reply tensorflow/python/keras/saving/save_test.py]: I'm not sure what you mean. This import is there so that the test can construct `pathlib.Path`s to check that the function works. We could move the import into the `test_save_load_tf_pathlib` function, though.

File: tensorflow/python/keras/utils/io_utils.py
Comment (ngc92) [reply tensorflow/python/keras/utils/io_utils.py]: I am pretty sure tensorflow or one of its dependencies will load the `os` module at some point, so the import should (as far as my understanding of the import process in python goes) be not much more than a simple dictionary lookup. 

You are right that the documentation looks rather ugly as is.

**Comment by ngc92** [comment]
I have added a commit that should add (partial) support for python 3.4/3.5. I have made the import to `os` unconditional, since I expect this to be fast (lookup, since this will be loaded elsewhere), and put the `pathlib` import for the 3.4/3.5 implementation into the function, so that this does not affect import times.

File: tensorflow/python/keras/saving/save_test.py
Comment (mihaimaruseac) [reply tensorflow/python/keras/saving/save_test.py]: Makes sense now, thank you

File: tensorflow/python/keras/utils/io_utils.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/utils/io_utils.py]: Please make sure docstring follows the style guide: one short phrase followed by empty line followed by rest of the docstring.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by alextp — COMMENTED** [review]
As tensorflow/api-owners this seems like a good change. @mihaimaruseac feel free to approve when you think it's ready

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by gbaned** [comment]
@ngc92 Can you please address Ubuntu Sanity errors? Thanks!

**Comment by ngc92** [comment]
I found that tensorflow offers a public API function `tf.compat.path_to_str` which mostly does the same thing my function does, only completely bypassing the python standard library. 

Should I remove my function in favour of the function that is already there?
I think the `compat` implementation does not support `pathlib` for python versions older than the fspath protocol, though (so 3.4 and 3.5).

The implementation there is simply
```
if hasattr(path, '__fspath__'):
    path = as_str_any(path.__fspath__())
return path
```

**Comment by mihaimaruseac** [comment]
I think that if we find a difference between the two function we can decide then which one to support and which one to deprecate.

If, however, both functions behave the same then we should use the existing one.

**Comment by ngc92** [comment]
there should only be a difference for python 3.4/3.5, because the `__fspath__()` attribute does not exist there.

apart from that, the question is whether it is cleaner to do things with python standard library functions instead of recoding them, i.e. the relevant documentation is
```
os.fspath(path)
    Return the file system representation of the path.
    If str or bytes is passed in, it is returned unchanged. Otherwise __fspath__() is called and its value is returned as long as it is a str or bytes object. In all other cases, TypeError is raised.
    New in version 3.6.

class os.PathLike
    An abstract base class for objects representing a file system path, e.g. pathlib.PurePath.
    New in version 3.6.
    abstractmethod __fspath__()
        Return the file system path representation of the object.
        The method should only return a str or bytes object, with the preference being for str.
```

The only problem is python 3.5 compatibility.

**Comment by mihaimaruseac** [comment]
Let's keep both for now as we still release 3.5 pips. Once we can deprecate python 3.5 pips we will deduplicate.

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**
